aptos-move/block-executor/src/scheduler_v2.rs

```diff
@@ -66,3 +432,1416 @@ impl ExecutionQueueManager {
+                self.queueing_commits_lock.arm();
+            }
+        }
+
+        if incarnation == 0 {
+            self.try_increase_executed_once_max_idx(txn_idx);
+        }
+
+        // Handle recursive propagation of add / remove stall.
+        self.propagate(stall_propagation_queue)?;
+
+        Ok(())
+    }
+
+    /// Sets the scheduler to a halted state.
+    ///
+    /// This is an atomic operation. Once halted, `[SchedulerV2::is_halted]` will return `true`,
+    /// and `[SchedulerV2::next_task]` will start returning `TaskKind::Done` once the
+    /// `post_commit_processing_queue` is empty.
+    /// This is typically used to signal an abnormal termination or a need to stop processing
+    /// transactions beyond a certain index.
+    ///
+    /// Returns `true` if this call actually changed the state from not-halted to halted,
+    /// `false` if it was already halted.
+    pub(crate) fn halt(&self) -> bool {
+        // TODO(BlockSTMv2): Notify waiting workers when supported.
+        !self.is_halted.swap(true, Ordering::SeqCst)
+    }
+
+    /// Checks if the scheduler is globally halted or if a specific transaction incarnation
+    /// has been aborted.
+    ///
+    /// This is used by workers during transaction execution to determine if they should
+    /// stop processing early.
+    ///
+    /// - Returns `true` if `[SchedulerV2::is_halted]` is true.
+    /// - If not globally halted and `incarnation == 0`, returns `false` (0-th incarnation
+    ///   is never aborted early to ensure its speculative writes are produced).
+    /// - Otherwise (not halted, `incarnation > 0`), returns the result of
+    ///   `[ExecutionStatuses::already_started_abort]`.
+    pub(crate) fn is_halted_or_aborted(&self, txn_idx: TxnIndex, incarnation: Incarnation) -> bool {
+        if self.is_halted() {
+            return true;
+        }
+
+        if incarnation == 0 {
+            // Never interrupt the 0-th incarnation due to an early abort to get the first output
+            // estimation (even if it is based on invalidated reads).
+            return false;
+        }
+
+        self.txn_statuses
+            .already_started_abort(txn_idx, incarnation)
+    }
+
+    // An interface for a particular transaction to be aborted directly, without abort manager.
+    // Abort manager is used with finish_execution API, which is a part of push invalidation
+    // flow and a core logic of BlockSTMv2. The direct_abort API below is targeted for other
+    // scenarios in which direct access to the abort functionality is helpful.
+    //
+    // This is currently used in three scenarios:
+    // (1) when processing dependencies (if the heuristics determines the txn is better to be
+    // speculatively aborted, e.g. since it is low priority and likely to be invalidated.
+    // Since the status must be Executing, we do not bother with propagating add / remove stall.
```

Comment:
This is safe because when we finish execution, we will propagate new invalidated set from this immediate re-execution, right?

Reply:
More fundamentally, the whole stall mechanism is an optimization and implemented as such, so if we omit adding (and removing the corresponding) stall, the correctness of the algorithm and implementation should be maintained.
finish execution on the status does not attempt to remove stall if previously the status was Executing as well, and propagation is purely local down the tree. 

That's technically, but the rationale for not adding and removing stall in this case is because it is already executing and much sooner we can expect the finish_execution (that would have to remove stall) then in aborted and other cases where the txn may not even be rescheduled yet (in which cases the stalls are added, propagated, etc).